docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,281 @@
+
+## Requirements
+
+ * Support the already integrated container runtime: `docker` and `rkt`
+ * Support hypervisor-based container runtimes: `hyper`.
+
+The existing pod-level interface will remain as it is in the near future to
+ensure supports of all existing runtimes are continued. Meanwhile, we will
+work with all parties involved to switching to the proposed interface.
+
+
+## Container Runtime Interface
+
+The main idea of this proposal is to adopt an imperative container-level
+interface, which allows Kubelet to directly control the lifecycles of the
+containers.
+
+Pod is composed of a group of containers in an isolated environment with
+resource constraints. In Kubernetes, pod is also the smallest schedulable unit.
+After a pod has been scheduled to the node, Kubelet will create the environment
+for the pod, and add/update/remove containers in that environment to meet the
+Pod Spec. To distinguish between the environment and the pod as a whole, we
+will call the pod environment **PodSandbox.**
+
+The container runtimes may interpret the PodSandBox concept differently based
+on how it operates internally. For runtimes relying on hypervisor, sandbox
+represents a virtual machine naturally. For others, it can be Linux namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Compute resource specifications**: A PodSandbox should implement pod-level
+   resource demands and restrictions.
+
+*NOTE: The resource specification does not include externalized costs to
+container setup that are not currently trackable as Pod constraints, e.g.,
+filesystem setup, container image pulling, etc.*
+
+A container in a PodSandbox maps to an application in the Pod Spec. For Linux
+containers, they are expected to share at least network and IPC namespaces,
+with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
+
+
+Below is an example of the proposed interfaces.
+
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Status(id string) PodSandboxStatus
+}
+
+// ContainerRuntime contains basic operations for containers.
+type ContainerRuntime interface {
+    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, PodSandboxID string) (string, error)
+    Start(id string) error
+    Stop(id string, timeout int) error
+    Remove(id string) error
+    List(filter ContainerFilter) ([]ContainerListItem, error)
+    Status(id string) (ContainerStatus, error)
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
+    List() ([]Image, error)
+    Pull(image ImageSpec, auth AuthConfig) error
```

Comment:
The intention is for us to create a cgroup for the pod prior to pulling any
images for any container used in that pod and starting them.

https://github.com/kubernetes/kubernetes/pull/26751#discussion_r68434206

Since we are exploring the ability to have a remote runtime for the Kubelet
here:

https://github.com/kubernetes/kubernetes/pull/17048

This change would allow the OCI runtime to explore the ability to charge
image pulling to the pod using tools like skopeo, and if/when we add blkio
support, we can account for the image pulling appropriately to the pod
rather than the node without needing to change the Kubelet remote API.

Runtimes that want to innovate in this space are then free to pursue it
from the start without requiring any version changes etc.

On Friday, June 24, 2016, Pengfei Ni <notifications@github.com> wrote:

> In docs/proposals/container-runtime-interface-v1.md
> <https://github.com/kubernetes/kubernetes/pull/25899#discussion_r68479985>
> :
>
> > +
> > +// ContainerRuntime contains basic operations for containers.
> > +type ContainerRuntime interface {
> > +    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, PodSandboxID string) (string, error)
> > +    Start(id string) error
> > +    Stop(id string, timeout int) error
> > +    Remove(id string) error
> > +    List(filter ContainerFilter) ([]ContainerListItem, error)
> > +    Status(id string) (ContainerStatus, error)
> > +	Exec(id string, cmd []string, streamOpts StreamOptions) error
> > +}
> > +
> > +// ImageOperations contains image-related operations.
> > +type ImageOperations interface {
> > +    List() ([]Image, error)
> > +    Pull(image ImageSpec, auth AuthConfig) error
>
> I would like to be able to charge image pulling to a pod in the future. I
> suspect I need the pod sandbox as input.
>
> @derekwaynecarr <https://github.com/derekwaynecarr> Could you explain the
> usecases of pulling image to a pod? And if image has not beed pulled, what
> image could be used to create the pod?
>
> I suppose this is not the use-case of docker, hyper or oci-runtime (they
> all needs the image before creating a pod). So I agree with @yujuhong
> <https://github.com/yujuhong> to add this later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25899/files/b5cd3a7f3c707d5542f247d77da44052cf386b25#r68479985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF8dbDTfly1qrotIKywvncsjnLWWW9mmks5qPHdAgaJpZM4Iil-->
> .
>